deepspeed/module_inject/replace_module.py

```diff
@@ -398,16 +399,104 @@ def _replace(child, name, conv_linear_layer):
+                        new_bias, prepare_tp_fused_qkvw(module_str, child.bias.data, mp_size)).to(
+                            get_accelerator().current_device_name())
+                else:
+                    data = mp_replace.copy(new_weight, child.weight.data)
+                    bias_data = None if child.bias is None else mp_replace.copy(new_bias, child.bias.data).to(
+                        get_accelerator().current_device_name())
+
                 setattr(child, "replaced", True)
                 return LinearLayer(weight=data.to(get_accelerator().current_device_name()), bias=bias_data)
 
+        def require_tp_fused_qkvw(name):
+            if mp_size == 1:
+                return False
+            fused_qkvw_name_list = ['qkv_proj', 'query_key_value']
+            for fused_name in fused_qkvw_name_list:
+                if fused_name in name:
+                    return True
+            return False
+
+        def prepare_tp_fused_qkvw(module_str, src, mp_size):
+            if src == None:
+                return
+            fused_type_dict = {
+                'CodeGenBlock': 'codegentype',
+                'BloomBlock': 'bloomtype',
+                'GLMBlock': 'glmtype',
+            }
+
+            def _codegen_type_transpose(input, mp_size, codegen_mp_num=4):
+                # codegen_mp_num defined in https://github.com/huggingface/transformers/blob/main/src/transformers/models/codegen/modeling_codegen.py
+                #TODO: assert num_heads % (mp_size*codegen_mp_num) == 0
+                #input : [3*hidden_dim, hidden_dim](weight) or [3*hidden_dim](bias)
+
+                shape = input.shape
+                dst_shape = shape[0] // mp_size
+                num_mp_blocks = input.reshape(codegen_mp_num, shape[0] // codegen_mp_num, shape[1])
+
+                #num_mp_blocks : [codegen_mp_num, 3*hidden_dim/codegen_mp_num, :]
+                src_split = list(torch.split(num_mp_blocks, num_mp_blocks.shape[1] // 3, dim=1))
+                src_split = [x.reshape(codegen_mp_num * mp_size, -1, shape[1]) for x in src_split]
+
+                qkv_split = [torch.split(src_s, num_mp_blocks.shape[1] // 3 // mp_size, dim=0) for src_s in src_split]
+
+                split_fuseqkv = [torch.cat([qkv_s[i] for qkv_s in qkv_split], dim=1) for i in range(len(qkv_split[0]))]
+                tp_fuseqkv_weight = torch.cat(split_fuseqkv, dim=0).reshape(shape[0], -1)
+
+                return tp_fuseqkv_weight[mp_replace.gpu_index * dst_shape:(mp_replace.gpu_index + 1) * dst_shape]
+
+            def _glm_type_transpose(input, mp_size):
+                #input : [3*hidden_dim, hidden_dim](weight) or [3*hidden_dim](bias)
+
+                shape = input.shape
+                dst_shape = shape[0] // mp_size
+                src_split = torch.split(input, shape[0] // 3, dim=0)
+
+                qkv_split = [torch.split(src_s, shape[0] // 3 // mp_size, dim=0) for src_s in src_split]
+                split_fusedqkv = [
```

Comment:
split_fusedqkv (cat operations) can be made as a separate method since there may be other variations using the same logic flow. @inkcherry ? 
Also changes look good, @molly-smith @RezaYazdaniAminabadi  could you please review and provide suggestions?

Reply:
Yes, I agree, and since we have the same logic for when using the TP for when the inference kernels are enabled then we can reuse this.